plugin.program.AML/resources/utils_kodi.py

```diff
@@ -0,0 +1,220 @@
+                                  
+        # At this point we are sure str_text is a unicode string.
+        log_text = u'AML DEBUG: ' + str_text
+        xbmc.log(log_text.encode('utf-8'), level = xbmc.LOGNOTICE)
+
+def log_verb(str_text):
+    if current_log_level >= LOG_VERB:
+        if isinstance(str_text, str): str_text = str_text.decode('utf-8')
+        log_text = u'AML VERB : ' + str_text
+        xbmc.log(log_text.encode('utf-8'), level = xbmc.LOGNOTICE)
+
+def log_info(str_text):
+    if current_log_level >= LOG_INFO:
+        if isinstance(str_text, str): str_text = str_text.decode('utf-8')
+        log_text = u'AML INFO : ' + str_text
+        xbmc.log(log_text.encode('utf-8'), level = xbmc.LOGNOTICE)
+
+def log_warning(str_text):
+    if current_log_level >= LOG_WARNING:
+        if isinstance(str_text, str): str_text = str_text.decode('utf-8')
+        log_text = u'AML WARN : ' + str_text
+        xbmc.log(log_text.encode('utf-8'), level = xbmc.LOGWARNING)
+
+def log_error(str_text):
+    if current_log_level >= LOG_ERROR:
+        if isinstance(str_text, str): str_text = str_text.decode('utf-8')
+        log_text = u'AML ERROR: ' + str_text
+        xbmc.log(log_text.encode('utf-8'), level = xbmc.LOGERROR)
+
+# -----------------------------------------------------------------------------
+# Kodi notifications and dialogs
+# -----------------------------------------------------------------------------
+#
+# Displays a modal dialog with an OK button. Dialog can have up to 3 rows of text, however first
+# row is multiline.
+# Call examples:
+#  1) ret = kodi_dialog_OK('Launch ROM?')
+#  2) ret = kodi_dialog_OK('Launch ROM?', title = 'AML - Launcher')
+#
+def kodi_dialog_OK(row1, row2='', row3='', title = 'Advanced MAME Launcher'):
+    dialog = xbmcgui.Dialog()
+    dialog.ok(title, row1, row2, row3)
+
+#
+# Returns True is YES was pressed, returns False if NO was pressed or dialog canceled.
+def kodi_dialog_yesno(row1, row2='', row3='', title = 'Advanced MAME Launcher'):
+    dialog = xbmcgui.Dialog()
+    ret = dialog.yesno(title, row1, row2, row3)
+
+    return ret
+
+#
+# Displays a small box in the bottom right corner
+#
+def kodi_notify(text, title = 'Advanced MAME Launcher', time = 5000):
+    dialog = xbmcgui.Dialog()
+    dialog.notification(title, text, xbmcgui.NOTIFICATION_INFO, time)
+
+def kodi_notify_warn(text, title = 'Advanced MAME Launcher warning', time = 7000):
+    dialog = xbmcgui.Dialog()
+    dialog.notification(title, text, xbmcgui.NOTIFICATION_WARNING, time)
+
+#
+# Do not use this function much because it is the same icon as when Python fails, and that may confuse the user.
+#
+def kodi_notify_error(text, title = 'Advanced MAME Launcher error', time = 7000):
+    dialog = xbmcgui.Dialog()
+    dialog.notification(title, text, xbmcgui.NOTIFICATION_ERROR, time)
+
+def kodi_busydialog_ON():
+    xbmc.executebuiltin('ActivateWindow(busydialog)')
```

Comment:
remove, it has been removed from Leia and results in noop

Reply:
https://github.com/xbmc/xbmc/pull/13958

Reply:
I need some help with this. In the Advanced Launchers is it very important to sometimes block the GUI when writing database JSON files to prevent running multiple instances of the addon when writing files. In old Advanced Launcher there was a very annoying bug due to the addon concurrency that cause database corruptions from time to time. The solution I found in the Advanced Launcher was to block the GUI. However, I read in the PRs you mentioned that FernetMenta wants to deprecate the BusyDialogs (he even mentions that addons should never block the GUI, if I understood correctly).

If I want to block the GUI for a time to avoid addon concurrency, how I should do it?

Reply:
The problem was that the busydialog can only be used once at a time. If multiple addons or services use it at the same time it can cause all sorts of strange behaviours, including crashes.

My suggestion is to at least create a new method somewhere in your addon where you apply a different logic depending on the Kodi version. If the user is running Kodi 18+ use `DialogNoCancel` instead. I thought it was exposed to the python API already but it looks like it is not. You can still use builtin functions to activate and deactivate the window:
```
ActivateWindow(window[,dir,return])
ReplaceWindow(window,dir)
```
Window ID is here: https://github.com/xbmc/xbmc/pull/13958/files#diff-580c97e14ab073901962f2ba5f7dfc55R108

For krypton your only option (apart from creating custom window) is to still use busydialog. 
The python API might change till v18 is final regarding the use of the this new window.

Reply:
OK, I understand FernetMenta motivations now. AEL uses BusyDialog() quite a lot. However, in AML I use DialogProgress() to block the UI. The reason is that BusyDialog() seems to introduce some delay but DialogProgress() is much faster, at least with Estuary skin (the delay may be skin indiced, I don't know).

In any case, I have removed all the BusyDialog()-related functions and will use DialogProgress() when I need to block the UI to avoid database corruption due to concurrency.